numpy/lib/shape_base.py

```diff
@@ -107,7 +107,7 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
             res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
             outarr[tuple(ind)] = res
             k += 1
-        return outarr
+        return arr.__array_wrap__(outarr)
     else:
         Ntot = product(outshape)
         holdshape = outshape
```

Comment:
Somewhat of an aside, but suppose one uses `func1d=np.mean`, then at least for astropy's `Quantity` class, what one gets is an `Quantity` array scalar as output. For any array scalar, `np.isscalar` returns `False`, so one ends up in this branch, but with an object that has `shape=()` and for which trying `len(a)` gives a `TypeError`:
```
a = np.array(1)
a.shape
# ()
len(a)
# TypeError(...)
```
To catch this, one has to make this branch aware of the possibility that `res.shape == ()`.